### ISSUE TITLE
Redirect: adding a redirect (301 or 302) option in ingressroute / httpproxy

### ISSUE THREAD

**Issue by vincentdeity** [projectcontour/contour#1627]
**Describe the solution you'd like**
The ingressroute / httpproxy should be able to support a match and then redirect with either a 301 or 302 to an http or https destination

Something like:
```
spec:
  routes:
  - match: /
    services:
    - name: awesome-service-01
      port: 1000
   - match: /redirectpath
    redirect:
      method: 301
      destination: https://externalawesomewebsite.com
  virtualhost:
    fqdn: awesomehost.tld
```

**Anything else you would like to add:**
The use case would be that for example if you have a website pointed to the ingress you can now also do redirects in contour. This would prevent you from having to spin up another container inside for the redirect (nginx for example).

**Environment:**

- Contour version: v1.0.0-beta.1
- Kubernetes version: (use `kubectl version`): v1.14.6-gke.1
- Kubernetes installer & version: N/A (Google)
- Cloud provider or hardware configuration: GKE
- OS (e.g. from `/etc/os-release`): GKE

**Comment by danramteke** [comment]
This would be nice for dealing with redirecting 'example.com` to `www.example.com` or vice-versa. Currently, I'm running an nginx container whose only job is to redirect all requests to `www.example.com`.

**Comment by davecheney** [comment]
Thank you for your feature request. I think I understand the use case. At the moment we're deep in feature freeze in preparation for Contour 1.0. I've assigned this to the backlog for prioritisation for a future release.

**Comment by PeopleRange** [comment]
Also needed from an SEO standpoint with the http to https or root_uri to www.root_uri

**Comment by jpeach** [comment]
> On 24 Jan 2020, at 2:59 am, PeopleRange <notifications@github.com> wrote:
> 
> Also needed from an SEO standpoint with the http to https or root_uri to www.root_uri

When TLS is enabled on a root HTTPProxy, the http -> https redirect is automatically provisioned.

https://projectcontour.io/docs/v1.1.0/httpproxy/#tls

**Comment by stevesloka** [comment]
A common use-case for this is redirecting `www.stevesloka.com` --> `stevesloka.com` which isn't the same as the tls redirect @jpeach

**Comment by PeopleRange** [comment]
@jpeach Thank You for the link to the documentation. I already have this bit implemented my experience comes from Nginx which common practice especially if using hsts headers is to 301 redirects to http then 301 redirect to the canonical domain (If Any).

**Comment by bgagnon** [comment]
We have another use case here for Intranet websites -- users often rely on the DNS search path so they type "foo" instead of "foo.company.org". A 3xx redirect from the short form to the long form is often necessary. Being able to do this as the ingress proxy layer is more convenient than implementing this code in every web app / backend service.

**Comment by jpeach** [comment]
From https://github.com/projectcontour/contour/issues/3014, some kinds of redirect will want to strip the query string. The example from that issue is:
```
spec:
  virtualhost: foo.example.com
  routes:
    - redirect:
        destination: bar.example.com  # required
        include_query_string: false      # default 
```

**Comment by skriss** [comment]
ref. https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route_components.proto#envoy-api-msg-route-redirectaction

**Comment by anthonyrea** [comment]
We require the same functionality to prevent the creation of a Pod specifically for a 301 Location redirect. Is there any update on feature prioritization?

**Comment by stevesloka** [comment]
Support for this got added in #3555 which adds DAG/Envoy support for direct responses. I think if we added a `RedirectPolicy` to the route that would make sense as proposed originally.

**Comment by youngnick** [comment]
Yes, I think that adding something like a `RedirectPolicy` as an option in HTTPProxy would be great. The next step here is a design, but we'll need to review the prioritization of this issue in a triage meeting soon.

**Comment by MichaelOgunsanmi** [comment]
Is the redirection described above now possible using contour? I have a similar use case as described by the OP and I have not been able to get a working solution based on the documentation available at https://projectcontour.io/docs/v1.16.0/config/external-service-routing/

**Comment by youngnick** [comment]
Hi @MichaelOgunsanmi, sorry but we haven't finished building this support out yet. #3555 added internal support for doing something similar to this (returning a 503 if the service is invalid), but we haven't added the ability to request a redirect on a specific route yet. We will need to be able to do this for the in-development upstream Gateway API, and we will backport similar functionality to HTTPProxy. 

But, we haven't finished that yet, sorry.

**Comment by xaleeks** [comment]
@youngnick  do we need this implemented in GW API upstream first or can we do this w/ HTTPProxy in parallel?  This feature would be good to have for both I think.  Maybe also add the ability to just set a prefix redirect, keeping the hostname unchanged.  Supporting regex on the incoming request would be a nice to have.

**Comment by skriss** [comment]
We've already got it for Gateway API: https://github.com/projectcontour/contour/pull/4123

**Comment by xaleeks** [comment]
Great, thanks Steve

**Comment by youngnick** [comment]
So the current TODO here is to backport the Gateway API redirect functionality to HTTPProxy somehow. We'll need an API design for it, although if we can reuse a lot of the Gateway API config, it should be able to be small.
